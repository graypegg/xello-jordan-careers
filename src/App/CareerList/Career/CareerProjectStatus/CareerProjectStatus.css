.CareerProjectStatus__wrapper {
  display: inline-block;
  background: #ccc;
  font-weight: bold;
  font-size: 1rem;
  margin-left: 20px;
}

.CareerProjectStatus__wrapper select {
  appearance: none;
  background: transparent;
  color: #FFF;
  font-size: 0.9rem;
  border: 0;
  border-radius: 0;
  padding: 14px 15px;
  text-align: center;
}

.CareerProjectStatus__wrapper--not-started {
  background: var(--dark-gradient-soft);
}

.CareerProjectStatus__wrapper--in-progress {
  background: var(--light-gradient-soft);
}

.CareerProjectStatus__wrapper--complete {
  background: var(--primary-gradient-soft);
  border-color: transparent;
}