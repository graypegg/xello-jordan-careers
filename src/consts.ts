export const enum ImageMode {
  High,
  Low
} 