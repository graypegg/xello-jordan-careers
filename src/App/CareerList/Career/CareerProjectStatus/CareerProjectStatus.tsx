import { ICareer } from '../../../../types'
import { EStatus } from '../../../../consts';
import * as React from 'react'

import './CareerProjectStatus.css'

interface ICareerProjectStatusProps {
  career: ICareer,
  onChange: (newCareer: ICareer) => void 
}

function getStatus (career: ICareer): EStatus {
  if (career.meta) {
    return career.meta.status ? career.meta.status : EStatus.NotStarted
  }
  return EStatus.NotStarted
}

function newCareerStatus (career: ICareer, status: EStatus): ICareer {
  return {
    ...career,
    meta: {
      ...career.meta,
      status
    }
  }
}

const applyOnChangeFactory = (props: ICareerProjectStatusProps, status: EStatus): ((event: React.FormEvent<HTMLSelectElement>) => void) => {
  return ((event) => {
    props.onChange(
      newCareerStatus(props.career, event.currentTarget.value as EStatus)
    )
  })
}

function getClassSuffix (career: ICareer): string {
  switch (career.meta ? career.meta.status : null) {
    case EStatus.NotStarted: return '--not-started'
    case EStatus.InProgress: return '--in-progress'
    case EStatus.Complete: return '--complete'
    default: return '--not-started'
  }
}

function CareerProjectStatus (props: ICareerProjectStatusProps): JSX.Element {
  return (
    <div className={"CareerProjectStatus__wrapper CareerProjectStatus__wrapper" + getClassSuffix(props.career)}>
      <select className="CareerProjectStatus__options" value={getStatus(props.career)} onChange={ applyOnChangeFactory(props, EStatus.Complete) }>
        { Object.keys(EStatus).map((statusKey: string) => (
          <option
            key={statusKey}
            value={EStatus[statusKey]}>
            {EStatus[statusKey]}
          </option>
        )) }
      </select>
    </div>
  )
}

export default CareerProjectStatus