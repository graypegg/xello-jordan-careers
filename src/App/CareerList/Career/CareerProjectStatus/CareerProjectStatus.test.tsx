import { ICareer } from '../../../../types'

import * as React from 'react'
import { shallow } from 'enzyme'
import CareerProjectStatus from './CareerProjectStatus';
import { EStatus } from '../../../../consts';

const career: ICareer = {
  title: 'Test',
  description: 'Test',
  notes: ['a', 'b'],
  image: 'TestImage',
  meta: {
    status: EStatus.NotStarted
  },
  id: 1
}

const onChangeMock = (_: ICareer) => { /* */ }

it('renders without crashing', () => {
  expect(
    shallow(<CareerProjectStatus career={career} onChange={onChangeMock} />)
  ).toMatchSnapshot();
})

Object.keys(EStatus).forEach((statusKey) => {
  it(`should display ${EStatus[statusKey]}`, () => {
    const newCareer = {
      ...career,
      meta: {
        ...career.meta,
        status: EStatus[statusKey]
      }
    }
    const wrapper = shallow(<CareerProjectStatus career={newCareer} onChange={onChangeMock} />)
    expect(wrapper.find('select').prop('value')).toBe(EStatus[statusKey])
  })

  it(`should apply proper class to wrapper for ${EStatus[statusKey]}`, () => {
    const newCareer = {
      ...career,
      meta: {
        ...career.meta,
        status: EStatus[statusKey]
      }
    }
    const wrapper = shallow(<CareerProjectStatus career={newCareer} onChange={onChangeMock} />)
    expect(wrapper.hasClass(`CareerProjectStatus__wrapper--${ EStatus[statusKey].toLowerCase().replace(' ', '-') }`)).toBe(true)
  })

  it(`should dispatch props.onChange when changed to ${EStatus[statusKey]}`, () => {
    const nextCareer = {
      ...career,
      meta: {
        ...career.meta,
        status: EStatus[statusKey]
      }
    }
    const onChangeMockSpied = jest.fn().mockImplementation(onChangeMock)
    const wrapper = shallow(<CareerProjectStatus career={career} onChange={onChangeMockSpied} />)
    wrapper.find('select').simulate('change', { currentTarget: { value: EStatus[statusKey] } })
    expect(onChangeMockSpied).toBeCalledWith(nextCareer)
  })
})


