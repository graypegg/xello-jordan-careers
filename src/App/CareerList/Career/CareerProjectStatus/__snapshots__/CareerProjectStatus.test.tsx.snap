// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders without crashing 1`] = `
ShallowWrapper {
  "length": 1,
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <CareerProjectStatus
    career={
      Object {
        "description": "Test",
        "id": 1,
        "image": "TestImage",
        "meta": Object {
          "status": "Not Started",
        },
        "notes": Array [
          "a",
          "b",
        ],
        "title": "Test",
      }
    }
    onChange={[Function]}
  />,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": undefined,
    "nodeType": "host",
    "props": Object {
      "children": <select
        className="CareerProjectStatus__options"
        onChange={[Function]}
        value="Not Started"
      >
        <option
          value="Not Started"
        >
          Not Started
        </option>
        <option
          value="In Progress"
        >
          In Progress
        </option>
        <option
          value="Complete"
        >
          Complete
        </option>
      </select>,
      "className": "CareerProjectStatus__wrapper CareerProjectStatus__wrapper--not-started",
    },
    "ref": null,
    "rendered": Object {
      "instance": null,
      "key": undefined,
      "nodeType": "host",
      "props": Object {
        "children": Array [
          <option
            value="Not Started"
          >
            Not Started
          </option>,
          <option
            value="In Progress"
          >
            In Progress
          </option>,
          <option
            value="Complete"
          >
            Complete
          </option>,
        ],
        "className": "CareerProjectStatus__options",
        "onChange": [Function],
        "value": "Not Started",
      },
      "ref": null,
      "rendered": Array [
        Object {
          "instance": null,
          "key": "NotStarted",
          "nodeType": "host",
          "props": Object {
            "children": "Not Started",
            "value": "Not Started",
          },
          "ref": null,
          "rendered": "Not Started",
          "type": "option",
        },
        Object {
          "instance": null,
          "key": "InProgress",
          "nodeType": "host",
          "props": Object {
            "children": "In Progress",
            "value": "In Progress",
          },
          "ref": null,
          "rendered": "In Progress",
          "type": "option",
        },
        Object {
          "instance": null,
          "key": "Complete",
          "nodeType": "host",
          "props": Object {
            "children": "Complete",
            "value": "Complete",
          },
          "ref": null,
          "rendered": "Complete",
          "type": "option",
        },
      ],
      "type": "select",
    },
    "type": "div",
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": undefined,
      "nodeType": "host",
      "props": Object {
        "children": <select
          className="CareerProjectStatus__options"
          onChange={[Function]}
          value="Not Started"
        >
          <option
            value="Not Started"
          >
            Not Started
          </option>
          <option
            value="In Progress"
          >
            In Progress
          </option>
          <option
            value="Complete"
          >
            Complete
          </option>
        </select>,
        "className": "CareerProjectStatus__wrapper CareerProjectStatus__wrapper--not-started",
      },
      "ref": null,
      "rendered": Object {
        "instance": null,
        "key": undefined,
        "nodeType": "host",
        "props": Object {
          "children": Array [
            <option
              value="Not Started"
            >
              Not Started
            </option>,
            <option
              value="In Progress"
            >
              In Progress
            </option>,
            <option
              value="Complete"
            >
              Complete
            </option>,
          ],
          "className": "CareerProjectStatus__options",
          "onChange": [Function],
          "value": "Not Started",
        },
        "ref": null,
        "rendered": Array [
          Object {
            "instance": null,
            "key": "NotStarted",
            "nodeType": "host",
            "props": Object {
              "children": "Not Started",
              "value": "Not Started",
            },
            "ref": null,
            "rendered": "Not Started",
            "type": "option",
          },
          Object {
            "instance": null,
            "key": "InProgress",
            "nodeType": "host",
            "props": Object {
              "children": "In Progress",
              "value": "In Progress",
            },
            "ref": null,
            "rendered": "In Progress",
            "type": "option",
          },
          Object {
            "instance": null,
            "key": "Complete",
            "nodeType": "host",
            "props": Object {
              "children": "Complete",
              "value": "Complete",
            },
            "ref": null,
            "rendered": "Complete",
            "type": "option",
          },
        ],
        "type": "select",
      },
      "type": "div",
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
      },
    },
  },
}
`;
